pkg/controller/volume/persistentvolume/util/util.go

```diff
@@ -238,7 +248,7 @@ func FindMatchingVolume(
 			return volume, nil
 		}
 
-		if node == nil && delayBinding {
+		if node == nil && (delayBinding || IsDelayBindingProvisioning(claim)) {
```

Comment:
I'm curious about this issue. If the PVC has `AnnSelectedNode` set by the scheduler, it must be in delay binding mode, otherwise, it will not be handled by the scheduler. I don't think this is necessary to add this check. Did I miss something?

Reply:
@cofyc the `delayBinding` param in this place is computed as storageclass is `WaitForFirstConsumer` **AND** NOT `selected-node` annotated, see the logic before in pv_controller.go:`shouldDelayBinding`

Reply:
after being scheduled, PVC with delay binding mode will be annotated with `AnnSelectedNode`, then it is treated as non-delayBinding should be dynamic provisioned ASAP

Reply:
@msau42 's opinion is that pv_controller should NOT find a PV for those PVCs `IsDelayBindingProvisioning` besides waiting for the dynamic provisioner. and i agree with this.

Reply:
Thanks for the clarification! IIUC, when PVC has `AnnSelectedNode` set, it must be dynamically provisioned in the PV controller, so the `ctrl.shouldDelayBinding(claim)` returns `false`. However, `util.FindMatchingVolume` will search all PVs under this storage class. If there is a static PV created and dynamic provisioner didn't complete the creation of the volume, the PVC may be bound to a wrong PV. So you add this check to skip the rest code.

This LGTM, add some comments to clarify it?

Reply:
yeah, i do mean that.

Reply:
I have a idea to fix this issue in `syncUnboundClaim`.

- `ctrl.shouldDelayBinding(claim)` returns true always, if the volume binding of clas `WaitForFirstConsumer`
   - `FindMatchingVolume` will be passed `delayBinding=true` when the PVC has `AnnSelectedNode` set
   - I guess this function can be replaced by `pvutil.IsDelayBindingMode(claim, ctrl.classLister)` with this change
- fix `syncUnboundClaim` to dynamically provision the PVC when the PVC has `AnnSelectedNode` set

This will make the code clearer. What do you think? @yittg @msau42

Reply:
For a quick review, like this?
func (ctrl *PersistentVolumeController) shouldDelayBinding(claim *v1.PersistentVolumeClaim) (bool, error) {
	// If claim is in delay binding mode.
	return pvutil.IsDelayBindingMode(claim, ctrl.classLister)
}

Reply:
For a quick review like this?
```
func (ctrl *PersistentVolumeController) shouldDelayBinding(claim *v1.PersistentVolumeClaim) (bool, error) {
	return pvutil.IsDelayBindingMode(claim, ctrl.classLister)
}
...
                  switch {
			case delayBinding && !IsDelayBindingProvisioning(claim):
				ctrl.eventRecorder.Event(claim, v1.EventTypeNormal, events.WaitForFirstConsumer, "waiting for first consumer to be created before binding")
			case v1helper.GetPersistentVolumeClaimClass(claim) != "":
				if err = ctrl.provisionClaim(claim); err != nil {
					return err
				}
				return nil
```

Reply:
yes

Reply:
but it do change the meaning of `delayBinding` which is passed through many functions,
and also some test_cases around it. so let's wait for @msau42 's suggestions? :)

Reply:
i looked into the history about function `shouldDelayBinding`, found that looks like this issue is introduced when adding the `AnnSelectedNode` check. this may be the right way to fix it :)

Reply:
sure

Reply:
I like the idea to skip `FindMatchingVolumes` if `AnnSelectedNode` is not set.  I'm not sure about completely skipping `FindMatchingVolume` for `WaitForFirstConsumer`; that will break static binding.

Reply:
@msau42 not completely i think, static binding can be applied when scheduling. Or do you mean the scenario that user create `WaitForFirstConsumer`  PVC without creating a pod consuming it.

Reply:
`FindMatchingVolumes` does two things:

- a) it searches pre-bound PV for the claim first, return when a PV is found
- b) when the node is given (in scheduler code path) or delayBinding is false (in PV controller code path when the class is not `WaitForFirstConsumer`), it searches a matching PV from all static PVs under given class.

In the current master, when `AnnSelectedNode` is set, `delayBinding` is set to true even if the class is `WaitForFirstConsumer`, this is not intended because it will search static PVs. What we need to do is to skip b). I suggest returning `delayBinding` as `true` always when the class is `WaitForFirstConsumer` and fix the issue in `syncUnboundClaim`.

The new workflow will be like this:

- PVC (no AnnSelectedNode and no WaitForFirstConsumer)
  - search pre-bound PV or a matching PV (FindMatchingVolumes)
  - if not found, it will try dynamic provisioning (`delayBinding && !IsDelayBindingProvisioning(claim)` == `false`)
- PVC (AnnSelectedNode and no WaitForFirstConsumer)
  - not possible because `AnnSelectedNode` is set by the scheduler
  - If it is set by the user, it works the same as the PVC without `AnnSelectedNode`. I think it's correct. (`delayBinding && !IsDelayBindingProvisioning(claim)` == `false`)
- PVC (no AnnSelectedNode and WaitForFirstConsumer)
  - search pre-bound PV (FindMatchingVolumes, no matching PV will be searched because delayBinding is true)
  - if not found, wait for the scheduler because the delayBinding is true and no  AnnSelectedNode set (`delayBinding && !IsDelayBindingProvisioning(claim)` == `true`)
- PVC (AnnSelectedNode and WaitForFirstConsumer)
  - search pre-bound PV (FindMatchingVolumes, no matching PV will be searched because delayBinding is true)
  - if not found, it will try dynamic provisioning (`delayBinding && !IsDelayBindingProvisioning(claim)` == `false`)

Reply:
@cofyc great clarity, maybe a little typo: it should be `false` here:
> In the current master, when `AnnSelectedNode` is set, `delayBinding` is set to **true** even if the class is `WaitForFirstConsumer`

Reply:
Fixed. Thanks!

Reply:
@yittg can you update to what I suggested? we can see if it breaks anything

Reply:
Ok, it's on the way